lib/ruby_lsp/extensions/hover.rb

```diff
@@ -0,0 +1,30 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RubyLsp
+  module Extensions
+    class Hover
+      extend T::Helpers
+
+      abstract!
+
+      class << self
+        extend T::Sig
+
+        sig { returns(T::Array[T.class_of(Extensions::Hover)]) }
+        def hooks
+          @hooks ||= T.let([], T.nilable(T::Array[T.class_of(Extensions::Hover)]))
+        end
+
+        sig { params(child_class: T.class_of(Extensions::Hover)).void }
+        def inherited(child_class)
+          hooks << child_class
+          super
+        end
+
+        sig { abstract.params(target: SyntaxTree::Node).returns(T.nilable(String)) }
+        def run(target); end
```

Comment:
What if the provider needs to reference other nodes in the document, or even other documents to get its job done?

Also, why a class method? Instance methods are easier to work with and the implementation might want to hold some kind of state in its implementation. If `run` was an instance method, the provider could accept a document, etc, state in its constructor, for example.

This was my biggest regret of the initial Tapioca compiler design, that I'd made the `decorate` method a class method.

Reply:
> About the class method, there's no big reason for it. Even if you wanted state, you could still do
> 
> ```ruby
> def self.run(target)
>   new(target).do_something_else
> end
> ```
> 
> But I'm okay with changing to an instance method.

You can do that, but if we are providing a base class for the hover provider, then we might as well make the base provider have an `initialize` method that accepts a `target` parameter and stores it in a `target` attr reader, so that the provider just needs to subclass and implement a `run` method without any parameters. I am suggesting that that would make it simpler to implement a provider than implementing a stateless class method.